Add OAuth and Multi-Record Query for SNOW

n3pjk opened the pull request:
##### SUMMARY
Add snow_record_find module to retrieve multiple records based on a query.

Add OAuth authentication support to snow_record and snow_record_find.

Fixes
- #58311  
- #58402

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
- snow_record.py
- snow_record_find.py

##### ADDITIONAL INFORMATION
If you use sso/ADFS for authenticating to SNOW, you will receive a 401 unauthorized response when trying to access using Ansible.  The solution is to create a user in SNOW that is not tied to ADFS.  Give the user OAuth access, with client Id and a client secret credentials.  Supply these creds in addition to the username and password to properly authenticate.

snow_record_find permits retrieval of multiple records based on a query.  The original module from Tim Rightnour has also been modified to support OAuth authentication.

```yaml
- name: Grab a user record using OAuth
  snow_record:
    username: ansible_test
    password: my_password
    client_id: "1234567890abcdef1234567890abcdef"
    client_secret: "Password1!"
    instance: dev99999
    state: present
    number: 62826bf03710200044e0bfc8bcbe5df1
    table: sys_user
    lookup_field: sys_id

- name: Using OAuth, search for incident assigned to group, return specific fields
  snow_record_find:
    username: ansible_test
    password: my_password
    client_id: "1234567890abcdef1234567890abcdef"
    client_secret: "Password1!"
    instance: dev99999
    table: incident
    query:
      assignment_group: d625dccec0a8016700a222a0f7900d06
    return_fields:
      - number
      - opened_at
```

Commit:
Add SNOW OAuth Support and Multi-record Query

Service Now (SNOW) modules currently do not support OAuth, even though
it is provided in pysnow. This functionality is added and tested in
both 2.7 and 2.8.

snow_record only performs create, update and delete operations on a
single record. When processing ticket queues, it is necessary to have a
module that can retrieve multiple records. snow_record_find does this,
and has also been modified to support OAuth. I have obtained permission
from the original author, Tim Rightnour <thegarbledone@gmail.com>, to
include this module with the tested upgrade. Email can be furnished if
necessary.

Commit:
Add OAuth documentation to snow_record_find

Commit:
Fix lint and verification issue for PR 58410

Commit:
Fix E309 and E324 errors in PR 58410

Commit:
Fix E307, need advice on E309

n3pjk commented:
I am unsure what needs to be done to fix the E309 error.  I tried changing the version_added from 2.5 to 2.9, but that caused an E307 error, which I have now corrected.

Commit:
Fix E309 for PR 58410

garbled1 commented:
shipit

Akasurde commented on lib/ansible/modules/notification/snow_record.py:
```diff
@@ -35,6 +35,16 @@
         required: true
+    client_id:
+        description:
+            - Client ID generated by ServiceNow
```
```suggestion
            - Client ID generated by ServiceNow.
```

Akasurde commented on lib/ansible/modules/notification/snow_record.py:
```diff
@@ -35,6 +35,16 @@
+        version_added: 2.9
+    client_secret:
+        description:
+            - Client Secret associated with client id
```
```suggestion
            - Client Secret associated with client id.
```

Akasurde commented on lib/ansible/modules/notification/snow_record.py:
```diff
@@ -148,7 +170,7 @@
 import os
 import traceback
 
-from ansible.module_utils.basic import AnsibleModule, missing_required_lib
```
I would give this since this Ansible API which will be handling missing required libraries.

Akasurde commented on lib/ansible/modules/notification/snow_record.py:
```diff
@@ -326,6 +381,20 @@ def run_module():
 
 
+def updater(new_token):
+    print("OAuth token refreshed!")
```
`print` statements are not used while Ansible run, you may want to use `module.warn` instead.

Akasurde commented on lib/ansible/modules/notification/snow_record_find.py:
```diff
@@ -0,0 +1,361 @@
+options:
+    instance:
+        description:
+            - The service now instance name
```
```suggestion
            - The service now instance name.
```

Akasurde reviewed (changes requested):
Rest LGTM

Commit:
Changes recommended by akasurde

n3pjk commented:
I've taken @Akasurde recommendations and applied them to both snow_record and snow_record_find.  I added periods to all descriptions, added missing_required_lib, and removed the print statement, which was a cut and paste artifact.

Akasurde committed:
Refactor service now module space

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

Akasurde committed:
Refactor service now module space

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

Akasurde commented:
@n3pjk I added additional comments in https://github.com/n3pjk/ansible/pull/1 Could you please review and test them ? Thanks.

Commit:
Re-add instance, username and password documentation

Commit:
Fix instance to use params

Commit:
Fix documentation indents

Commit:
Fix documentation indents pt 2

Commit:
Fix result dict init and import missing_required_lib

Commit:
Fix data type mismatch in documentation

Commit:
Fix future boilerplate

Commit:
Remove doc_fragment overlap

n3pjk commented:
@Akasurde I've merged your changes in, and fixed up all issues except the 'A102 Remove since...' issue on snow_record.py  A 102 error looks like it is complaining about an interpreter line not referencing powershell, but one doesn't exist.  Any thoughts?

n3pjk commented:
@Akasurde I wasn't quite finished testing the refactor, and I just figured out where the A102 error was coming from.  The service_now.py module is missing the python interpreter line at the top.  This causes the module import to fail. I'll open a bug fix for it since this PR is closed.

Akasurde commented:
@n3pjk Oh ! I am extremely sorry. Please open a PR which will complete your refactoring and let me know. 
Thanks.

n3pjk commented:
@Akasurde No worries! Actually, this error made me realize that I needed to remove the python shebang from the modules. Your refactor did not have a shebang, but the modules did. That caused an inconsistency. It will be better because of it!  I'm chasing down a potential problem with our snow instance before submitting the bug report, in case there's something additional lurking.

n3pjk commented:
See Issue #59111
